lib/client.js

```diff
@@ -279,34 +269,26 @@ Object.assign(Client.prototype, {
       ) {
         return; // not expired, reuse existing connection
       }
-      const providerResult = this.connectionConfigProvider();
-      if (!providerResult.then) {
-        handleConnectionConfigProviderResult(providerResult);
-        return; // provider is synchronous, no need to return a promise
+      const providerResult = await this.connectionConfigProvider();
+      if (providerResult.expirationChecker) {
+        this.connectionConfigExpirationChecker =
+          providerResult.expirationChecker;
+        delete providerResult.expirationChecker;
```

Comment:
why not just set to undefined? AFAIR it's still considerably faster.

Reply:
I replaced the delete of `this.connectionConfigExpirationChecker` a couple of lines later into null, because it seems a bit strange for the the client's own attributes to be just coming and going.

But as for this delete, there is a catch. The MySQL2 driver prints this nasty message if the expirationChecker property is present in the config, even if its value is `undefined`:
`Ignoring invalid configuration option passed to Connection: expirationChecker. This is currently a warning, but in future versions of MySQL2, an error will be thrown if you pass an invalid configuration option to a Connection`

Reply:
Can you add comment with this explanation, so that noone removes it in the future?